File: api/src/routers/social_network/social.py
"""
Layer: Interface Adapters — Web (Routers)
Package: routers.social_network.social
Responsibility: HTTP-контроллеры профилей, друзей, сообществ, трендов, жалоб
и графа социальной сети. Мутации требуют авторизации.
"""
import logging
from typing import Optional

from fastapi import APIRouter, Depends, File, Form, HTTPException, UploadFile
from fastapi.responses import Response
from pydantic import BaseModel

from services.social_network_service import SocialNetworkService
from web.dependencies import get_current_user, get_optional_user

logger = logging.getLogger(__name__)

router = APIRouter(tags=["social_network"])

service = SocialNetworkService()


class UpdateProfileRequest(BaseModel):
    bio: Optional[str] = None
    avatar_key: Optional[str] = None
    contacts: Optional[dict] = None


class CreateCommunityRequest(BaseModel):
    name: str
    description: str = ""


class UpdateCommunityRequest(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None


class CreateComplaintRequest(BaseModel):
    target_type: str
    target_uid: str
    reason: str
    comment: str = ""


# ── Профиль / me ─────────────────────────────────────────────────────────────

@router.get("/social/me")
async def get_me(user: dict = Depends(get_current_user)):
    return service.get_me(user)


@router.put("/social/profile")
async def update_profile(
    req: UpdateProfileRequest,
    user: dict = Depends(get_current_user),
):
    result = service.update_profile(
        user.get("uid", ""),
        bio=req.bio,
        avatar_key=req.avatar_key,
        contacts=req.contacts,
    )
    if not result.get("success"):
        raise HTTPException(status_code=404, detail=result.get("error", "user_not_found"))
    return result


# ── Пользователи ─────────────────────────────────────────────────────────────

@router.get("/social/users/search")
async def search_users(q: str = "", limit: int = 20, user: Optional[dict] = Depends(get_optional_user)):
    return {"success": True, "users": service.search_users(q, limit=min(limit, 100))}


@router.get("/social/users/{uid}")
async def get_user(uid: str, user: Optional[dict] = Depends(get_optional_user)):
    profile = service.get_user(uid)
    if profile is None:
        raise HTTPException(status_code=404, detail="user_not_found")
    return {"success": True, "user": profile}


@router.get("/social/users/{uid}/profile")
async def get_user_profile(uid: str, user: Optional[dict] = Depends(get_optional_user)):
    profile = service.get_profile(uid, viewer_uid=(user or {}).get("uid"))
    if profile is None:
        raise HTTPException(status_code=404, detail="user_not_found")
    return {"success": True, "profile": profile}


# ── Друзья ───────────────────────────────────────────────────────────────────

@router.get("/social/friends")
async def list_friends(user: dict = Depends(get_current_user)):
    return {"success": True, "friends": service.list_friends(user.get("uid", ""))}


@router.post("/social/friends/{friend_uid}")
async def add_friend(friend_uid: str, user: dict = Depends(get_current_user)):
    result = service.add_friend(user.get("uid", ""), friend_uid)
    if not result.get("success"):
        raise HTTPException(status_code=400, detail=result.get("error", "add_friend_failed"))
    return result


@router.delete("/social/friends/{friend_uid}")
async def remove_friend(friend_uid: str, user: dict = Depends(get_current_user)):
    return service.remove_friend(user.get("uid", ""), friend_uid)


# ── Сообщества ───────────────────────────────────────────────────────────────

@router.post("/social/communities")
async def create_community(req: CreateCommunityRequest, user: dict = Depends(get_current_user)):
    name = req.name.strip()
    if not name:
        raise HTTPException(status_code=400, detail="empty_name")
    community = service.create_community(
        owner_uid=user.get("uid", ""), name=name, description=req.description.strip()
    )
    return {"success": True, "community": community}


@router.get("/social/communities")
async def list_communities(limit: int = 100, user: Optional[dict] = Depends(get_optional_user)):
    viewer = (user or {}).get("uid")
    return {
        "success": True,
        "communities": service.list_communities(limit=min(limit, 200), viewer_uid=viewer),
    }


@router.get("/social/communities/search")
async def search_communities(q: str = "", limit: int = 20, user: Optional[dict] = Depends(get_optional_user)):
    viewer = (user or {}).get("uid")
    return {
        "success": True,
        "communities": service.search_communities(q, limit=min(limit, 100), viewer_uid=viewer),
    }


@router.get("/social/communities/mine")
async def my_communities(user: dict = Depends(get_current_user)):
    return {
        "success": True,
        "communities": service.list_owned_communities(user.get("uid", "")),
    }


@router.get("/social/communities/{uid}")
async def get_community(uid: str, user: Optional[dict] = Depends(get_optional_user)):
    community = service.get_community(uid, viewer_uid=(user or {}).get("uid"))
    if community is None:
        raise HTTPException(status_code=404, detail="community_not_found")
    return {"success": True, "community": community}


@router.post("/social/communities/{uid}/join")
async def join_community(uid: str, user: dict = Depends(get_current_user)):
    result = service.join_community(user.get("uid", ""), uid)
    if not result.get("success"):
        raise HTTPException(status_code=404, detail=result.get("error", "community_not_found"))
    return result


@router.post("/social/communities/{uid}/leave")
async def leave_community(uid: str, user: dict = Depends(get_current_user)):
    return service.leave_community(user.get("uid", ""), uid)


@router.put("/social/communities/{uid}")
async def update_community(
    uid: str,
    req: UpdateCommunityRequest,
    user: dict = Depends(get_current_user),
):
    result = service.update_community(
        user.get("uid", ""), uid, name=req.name, description=req.description
    )
    if not result.get("success"):
        code = 403 if result.get("error") == "forbidden" else 404
        raise HTTPException(status_code=code, detail=result.get("error", "community_not_found"))
    return result


@router.delete("/social/communities/{uid}")
async def delete_community(uid: str, user: dict = Depends(get_current_user)):
    result = service.delete_community(user.get("uid", ""), uid)
    if not result.get("success"):
        code = 403 if result.get("error") == "forbidden" else 404
        raise HTTPException(status_code=code, detail=result.get("error", "community_not_found"))
    return result


# ── Тренды ───────────────────────────────────────────────────────────────────

@router.get("/social/trends")
async def get_trends(limit: int = 10):
    return service.get_trends(limit=min(limit, 50))


# ── Жалобы ───────────────────────────────────────────────────────────────────

@router.post("/social/complaints")
async def create_complaint(req: CreateComplaintRequest, user: dict = Depends(get_current_user)):
    if not req.reason.strip():
        raise HTTPException(status_code=400, detail="empty_reason")
    result = service.create_complaint(
        reporter_uid=user.get("uid", ""),
        target_type=req.target_type,
        target_uid=req.target_uid,
        reason=req.reason.strip(),
        comment=req.comment.strip(),
    )
    if not result.get("success"):
        raise HTTPException(status_code=400, detail=result.get("error", "bad_target_type"))
    return result


# ── Граф ─────────────────────────────────────────────────────────────────────

@router.get("/social/graph")
async def get_social_graph(user: Optional[dict] = Depends(get_optional_user)):
    uid = (user or {}).get("uid")
    if uid:
        return service.get_ego_graph(uid)
    return service.get_public_graph()


@router.get("/social/graph/user/{uid}")
async def get_user_graph(uid: str, user: Optional[dict] = Depends(get_optional_user)):
    result = service.get_user_graph(uid, viewer_uid=(user or {}).get("uid"))
    if not result.get("success"):
        raise HTTPException(status_code=404, detail=result.get("error", "user_not_found"))
    return result


# ── Изображения ──────────────────────────────────────────────────────────────

@router.post("/social/images")
async def upload_image(
    file: UploadFile = File(...),
    user: dict = Depends(get_current_user),
):
    """Загружает изображение для чата в S3, возвращает object_key."""
    data = await file.read()
    if not data:
        raise HTTPException(status_code=400, detail="Empty file")
    result = await service.upload_image(
        user.get("uid", ""),
        file.filename or "image",
        file.content_type or "",
        data,
    )
    if not result.get("success"):
        raise HTTPException(status_code=500, detail=result.get("error", "Upload failed"))
    return result


@router.get("/social/images/{object_key:path}")
async def get_image(object_key: str):
    """Отдаёт содержимое изображения из S3 с корректным content-type."""
    image = await service.get_image(object_key)
    if image is None:
        raise HTTPException(status_code=404, detail="Image not found")
    data, content_type = image
    return Response(content=data, media_type=content_type)


@router.delete("/social/images/{object_key:path}")
async def delete_image(object_key: str, user: dict = Depends(get_current_user)):
    """Удаляет изображение из S3."""
    ok = await service.delete_image(object_key)
    if not ok:
        raise HTTPException(status_code=404, detail="Image not found")
    return {"success": True}


# ── Вклад пользователя / уведомления / resolve ──────────────────────────────

@router.get("/social/users/{uid}/contributions")
async def get_contributions(uid: str, user: Optional[dict] = Depends(get_optional_user)):
    return {"success": True, "contributions": service.get_contributions(uid)}


@router.get("/social/notifications/unread-count")
async def get_unread_count(user: dict = Depends(get_current_user)):
    return {"success": True, "unread_count": service.get_unread_count(user.get("uid", ""))}


@router.get("/social/resolve/{uid}")
async def resolve_entity(uid: str, user: Optional[dict] = Depends(get_optional_user)):
    entity = service.resolve_entity(uid)
    if entity is None:
        raise HTTPException(status_code=404, detail="entity_not_found")
    return {"success": True, "entity": entity}

File: client/src/services/api/social.ts
import { fetchJson } from './http';

/** Событие, уведомляющее UI о том, что профиль текущего пользователя изменён.
 *  Обработчики (панель профиля, аватар в шапке) перечитывают свои данные. */
export const PROFILE_UPDATED_EVENT = 'km:profile-updated';

/** Событие запроса на открытие модального окна «Личный кабинет».
 *  Испускается из любого места UI (например, из панели «Профиль»). */
export const ACCOUNT_MODAL_EVENT = 'km:open-account-modal';

// ── Типы социальной сети ───────────────────────────────────────────────────

export interface SocialContacts {
    email?: string;
    telegram?: string;
    phone?: string;
    website?: string;
}

export interface SocialUserProfile {
    uid: string;
    login: string;
    nickname: string;
    bio?: string;
    avatar_key?: string;
    contacts?: SocialContacts;
    friend_count?: number;
    communities?: CommunitySummary[];
    is_friend?: boolean;
    contributions?: { article_count: number; block_count: number };
}

export interface CommunitySummary {
    uid: string;
    name: string;
    description: string;
    member_count: number;
    created_by_uid?: string;
    is_member?: boolean;
}

export interface Community extends CommunitySummary {
    members: SocialUserProfile[];
    is_member?: boolean;
    is_owner?: boolean;
}

export type ChatTargetType = 'article' | 'statement' | 'user' | 'community';

export interface ChatReference {
    uid: string;
    type: 'block' | 'article' | 'statement';
    label: string;
    block_type?: number;
    order?: number;
}

export interface ChatMessage {
    uid: string;
    target_type: ChatTargetType;
    target_uid: string;
    author_uid: string;
    author_nickname: string;
    author_login: string;
    text: string;
    parent_uid: string;
    like_count: number;
    reply_count: number;
    liked_by_me: boolean;
    created_at: number;
    references?: ChatReference[];
}

export interface ChatThread {
    success: boolean;
    messages: ChatMessage[];
    total: number;
}

// ── Стена профиля ───────────────────────────────────────────────────────────

export interface WallComment {
    uid: string;
    post_uid: string;
    author_uid: string;
    author_nickname: string;
    author_login: string;
    text: string;
    created_at: number;
}

export interface WallPost {
    uid: string;
    owner_uid: string;
    text: string;
    created_at: number;
    comment_count: number;
    comments: WallComment[];
}

export interface WallResponse {
    success: boolean;
    posts: WallPost[];
    total: number;
}

export async function getWall(uid: string): Promise<WallResponse> {
    return fetchJson(`/api/social/wall/${encodeURIComponent(uid)}`, { method: 'GET' });
}

export async function createWallPost(
    uid: string,
    text: string,
): Promise<{ success: boolean; post?: WallPost; error?: string }> {
    return fetchJson(`/api/social/wall/${encodeURIComponent(uid)}`, {
        method: 'POST',
        body: JSON.stringify({ text }),
    });
}

export async function addWallComment(
    postUid: string,
    text: string,
): Promise<{ success: boolean; comment?: WallComment; error?: string }> {
    return fetchJson(`/api/social/wall/posts/${encodeURIComponent(postUid)}/comments`, {
        method: 'POST',
        body: JSON.stringify({ text }),
    });
}

export interface NotificationItem {
    uid: string;
    type: string;
    target_type: string;
    target_uid: string;
    text: string;
    is_read: boolean;
    created_at: number;
}

export interface NotificationsResponse {
    success: boolean;
    notifications: NotificationItem[];
    unread_count: number;
}

export interface TrendItem {
    target_type: ChatTargetType;
    target_uid: string;
    label: string;
    count: number;
}

export interface TrendsResponse {
    success: boolean;
    by_comments: TrendItem[];
    by_likes: TrendItem[];
}

export interface SocialGraphNode {
    id: string;
    type: 'user' | 'community';
    label: string;
    login?: string;
    description?: string;
    avatar_key?: string;
    friend_count?: number;
    member_count?: number;
    is_me?: boolean;
    is_friend?: boolean;
    is_member?: boolean;
}

export interface SocialGraphEdge {
    source: string;
    target: string;
    type: 'friend' | 'member';
}

export interface SocialGraph {
    success: boolean;
    nodes: SocialGraphNode[];
    edges: SocialGraphEdge[];
}

export interface MeResponse {
    success: boolean;
    profile: SocialUserProfile;
    friends: SocialUserProfile[];
    communities: CommunitySummary[];
    notifications: NotificationItem[];
    unread_count: number;
}

// ── Профиль / me ────────────────────────────────────────────────────────────

export async function getMe(): Promise<MeResponse> {
    return fetchJson<MeResponse>('/api/social/me', { method: 'GET' });
}

export async function updateProfile(body: { bio?: string; avatar_key?: string; contacts?: SocialContacts }): Promise<{
    success: boolean;
    profile?: SocialUserProfile;
    error?: string;
}> {
    return fetchJson('/api/social/profile', { method: 'PUT', body: JSON.stringify(body) });
}

// ── Пользователи / друзья ───────────────────────────────────────────────────

export async function searchUsers(q: string, limit = 20): Promise<{ success: boolean; users: SocialUserProfile[] }> {
    const params = new URLSearchParams({ q, limit: String(limit) });
    return fetchJson(`/api/social/users/search?${params.toString()}`, { method: 'GET' });
}

export async function getUserProfile(uid: string): Promise<{ success: boolean; profile?: SocialUserProfile }> {
    return fetchJson(`/api/social/users/${encodeURIComponent(uid)}/profile`, { method: 'GET' });
}

export async function listFriends(): Promise<{ success: boolean; friends: SocialUserProfile[] }> {
    return fetchJson('/api/social/friends', { method: 'GET' });
}

export async function addFriend(uid: string): Promise<{ success: boolean; is_friend?: boolean; error?: string }> {
    return fetchJson(`/api/social/friends/${encodeURIComponent(uid)}`, { method: 'POST' });
}

export async function removeFriend(uid: string): Promise<{ success: boolean; is_friend?: boolean }> {
    return fetchJson(`/api/social/friends/${encodeURIComponent(uid)}`, { method: 'DELETE' });
}

// ── Сообщества ──────────────────────────────────────────────────────────────

export async function createCommunity(body: {
    name: string;
    description?: string;
}): Promise<{ success: boolean; community?: CommunitySummary; error?: string }> {
    return fetchJson('/api/social/communities', { method: 'POST', body: JSON.stringify(body) });
}

export async function listCommunities(limit = 100): Promise<{ success: boolean; communities: CommunitySummary[] }> {
    return fetchJson(`/api/social/communities?limit=${limit}`, { method: 'GET' });
}

export async function searchCommunities(q: string, limit = 20): Promise<{
    success: boolean;
    communities: CommunitySummary[];
}> {
    const params = new URLSearchParams({ q, limit: String(limit) });
    return fetchJson(`/api/social/communities/search?${params.toString()}`, { method: 'GET' });
}

export async function getMyCommunities(): Promise<{
    success: boolean;
    communities: CommunitySummary[];
}> {
    return fetchJson('/api/social/communities/mine', { method: 'GET' });
}

export async function getCommunity(uid: string): Promise<{ success: boolean; community?: Community }> {
    return fetchJson(`/api/social/communities/${encodeURIComponent(uid)}`, { method: 'GET' });
}

export async function joinCommunity(uid: string): Promise<{ success: boolean; is_member?: boolean; error?: string }> {
    return fetchJson(`/api/social/communities/${encodeURIComponent(uid)}/join`, { method: 'POST' });
}

export async function leaveCommunity(uid: string): Promise<{ success: boolean; is_member?: boolean }> {
    return fetchJson(`/api/social/communities/${encodeURIComponent(uid)}/leave`, { method: 'POST' });
}

// ── Чат ─────────────────────────────────────────────────────────────────────

export async function getChat(
    targetType: ChatTargetType,
    targetUid: string,
    before?: number,
    limit = 50,
): Promise<ChatThread> {
    const params = new URLSearchParams({ limit: String(limit) });
    if (before) params.set('before', String(before));
    return fetchJson(
        `/api/social/chat/${targetType}/${encodeURIComponent(targetUid)}?${params.toString()}`,
        { method: 'GET' },
    );
}

export async function sendMessage(
    targetType: ChatTargetType,
    targetUid: string,
    text: string,
    parentUid?: string,
    references?: { uid: string; type: 'block' | 'article' | 'statement' }[],
): Promise<{ success: boolean; message?: ChatMessage; error?: string }> {
    return fetchJson(`/api/social/chat/${targetType}/${encodeURIComponent(targetUid)}`, {
        method: 'POST',
        body: JSON.stringify({
            text,
            parent_uid: parentUid ?? null,
            references: references && references.length ? references : null,
        }),
    });
}

export async function toggleLike(messageUid: string): Promise<{
    success: boolean;
    liked?: boolean;
    like_count?: number;
    error?: string;
}> {
    return fetchJson(`/api/social/chat/messages/${encodeURIComponent(messageUid)}/like`, { method: 'POST' });
}

// ── Контакты (обфускация) ───────────────────────────────────────────────────

/**
 * Сервер отдаёт контакты в base64, чтобы массовые краулеры не могли собрать
 * их простым парсингом JSON/HTML. Расшифровка выполняется в браузере перед
 * отрисовкой.
 */
export function decodeContacts(contacts?: SocialContacts): SocialContacts {
    if (!contacts) return {};
    const decoded: SocialContacts = {};
    for (const [key, value] of Object.entries(contacts)) {
        if (!value) continue;
        let text = value;
        try {
            const binary = atob(value);
            const bytes = Uint8Array.from(binary, (ch) => ch.charCodeAt(0));
            text = new TextDecoder('utf-8').decode(bytes);
        } catch {
            // не base64 — оставляем как есть (устаревшие данные)
        }
        decoded[key as keyof SocialContacts] = text;
    }
    return decoded;
}

// ── Уведомления ─────────────────────────────────────────────────────────────

export async function getNotifications(): Promise<NotificationsResponse> {
    return fetchJson('/api/social/notifications', { method: 'GET' });
}

export async function markNotificationsRead(notificationUid?: string): Promise<{ success: boolean }> {
    return fetchJson('/api/social/notifications/read', {
        method: 'POST',
        body: JSON.stringify({ notification_uid: notificationUid ?? null }),
    });
}

// ── Тренды / жалобы / граф ──────────────────────────────────────────────────

export async function getTrends(limit = 10): Promise<TrendsResponse> {
    return fetchJson(`/api/social/trends?limit=${limit}`, { method: 'GET' });
}

export async function createComplaint(body: {
    target_type: ChatTargetType;
    target_uid: string;
    reason: string;
    comment?: string;
}): Promise<{ success: boolean; complaint_uid?: string; error?: string }> {
    return fetchJson('/api/social/complaints', { method: 'POST', body: JSON.stringify(body) });
}

export async function getSocialGraph(): Promise<SocialGraph> {
    return fetchJson('/api/social/graph', { method: 'GET' });
}

export async function getGraphUser(uid: string): Promise<SocialGraph> {
    return fetchJson(`/api/social/graph/user/${encodeURIComponent(uid)}`, { method: 'GET' });
}

// ── Изображения ─────────────────────────────────────────────────────────────

export interface UploadImageResult {
    success: boolean;
    object_key?: string;
    error?: string;
}

export async function uploadImage(file: File): Promise<UploadImageResult> {
    const fd = new FormData();
    fd.append('file', file);
    return fetchJson('/api/social/images', { method: 'POST', body: fd });
}

export function socialImageUrl(objectKey: string): string {
    return `/api/social/images/${objectKey}`;
}

export async function deleteSocialImage(objectKey: string): Promise<{ success: boolean }> {
    return fetchJson(`/api/social/images/${encodeURIComponent(objectKey)}`, { method: 'DELETE' });
}

// ── Вклад / уведомления / resolve ───────────────────────────────────────────

export async function getContributions(uid: string): Promise<{
    success: boolean;
    contributions?: { article_count: number; block_count: number };
}> {
    return fetchJson(`/api/social/users/${encodeURIComponent(uid)}/contributions`, { method: 'GET' });
}

export async function getUnreadCount(): Promise<{ success: boolean; unread_count: number }> {
    return fetchJson('/api/social/notifications/unread-count', { method: 'GET' });
}

export interface ResolvedEntity {
    uid: string;
    type: 'block' | 'article' | 'statement' | 'community' | 'user' | 'message';
    label: string;
    block_type?: number;
    order?: number;
    target_type?: string;
    data?: Record<string, string | boolean | number | null>;
}

export async function resolveEntity(uid: string): Promise<{ success: boolean; entity?: ResolvedEntity }> {
    return fetchJson(`/api/social/resolve/${encodeURIComponent(uid)}`, { method: 'GET' });
}

// ── Управление сообществами ─────────────────────────────────────────────────

export async function updateCommunity(uid: string, body: {
    name?: string;
    description?: string;
}): Promise<{ success: boolean; community?: Community; error?: string }> {
    return fetchJson(`/api/social/communities/${encodeURIComponent(uid)}`, {
        method: 'PUT',
        body: JSON.stringify(body),
    });
}

export async function deleteCommunity(uid: string): Promise<{ success: boolean; error?: string }> {
    return fetchJson(`/api/social/communities/${encodeURIComponent(uid)}`, { method: 'DELETE' });
}
